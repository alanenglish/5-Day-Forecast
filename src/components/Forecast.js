import React from 'react';
import { toast } from 'react-toastify';
import ForecastDetails from './ForecastDetails';
import Loading from './Loading';
import SearchBar from './SearchBar';
import fetchForecasts from '../utils/API';

class Forecast extends React.Component {
  constructor() {
    super();

    this.state = {
      forecasts: [],
      city: '',
      state: '',
      error: '',
      isLoading: false
    };
  }

  onSubmit = (woe) => {
    this.setState({
      isLoading: true,
      forecasts: []
    });

    this.fetchWeatherForecast(woe);
  }

  async fetchWeatherForecast(woe) {
    this.setState({ error: '' });
    const response = await fetchForecasts(woe);

    if (response.status !== 200) {
      this.setState({
        error: response.error,
        isLoading: false
      });
    } else {
      this.setState({
        forecasts: response.data.consolidated_weather.slice(0, 5),
        city: response.data.title,
        state: response.data.parent.title,
        isLoading: false
      });
    }
  }

  displayForecastItems = () => this.state.forecasts.map((forecast) => <ForecastDetails key={forecast.id} forecast={forecast} />);

  render() {
    const { isLoading, error, forecasts } = this.state;
    if (error) toast.error(error);
    return (
      <div>
        <SearchBar onSubmit={this.onSubmit} />
        { isLoading ?
          <Loading />
        :
          <div className={`forecast-container ${!forecasts.length ? 'empty-state__background-img' : ''}`}>
            { !this.state.forecasts.length &&
              <div className="empty-state">
                <p className="empty_state__message">
                  Please enter a WOE ID in order to display forecast. Try running through the index for your WOE's.
                </p>
                <em className="empty-state__drake">-Drake</em>
              </div>
            }
            { !!this.state.forecasts.length &&
              <div>
                <div className="location">
                  {this.state.city}, {this.state.state}
                </div>
                <ul className="list-inline forecast-list">
                  {this.displayForecastItems()}
                </ul>
              </div>
            }
          </div>
        }
      </div>
    );
  }
}

export default Forecast;
