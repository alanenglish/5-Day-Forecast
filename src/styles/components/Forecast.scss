.forecast-container {
  align-items: center;
  display: flex;
  height: 75vh;
  justify-content: center;
  width: 100vw;
}

.empty-state__background-img {
  background-image: url(/images/weather.jpg);
  background-size: cover;
}

.empty-state {
  text-align: center;

  .empty_state__message {
    font-family: 'Oswald';
    font-size: 34px;
  }

  .empty-state__drake {
    font-family: 'Yesteryear', cursive;
    font-size: 30px;
    font-weight: bold;
  }
}

.location {
  font-family: 'Oswald';
  font-size: 64px;
  font-weight: bold;
  text-align: center;
}

.forecast-list {
  margin-bottom: 75px;
  margin-top: $xl-size;
}

@media only screen and (max-width: 768px) {
    /* For mobile phones: */
    .forecast-container {
      height: 84vh;
    }

    .location {
      font-size: 44px;
    }

    .forecast-list {
      margin-bottom: 200px;
      margin-top: $xl-size;
    }

    .empty-state {
      .empty_state__message {
        font-size: 26px;
        padding: 0 $s-size;
      }

      .empty-state__drake {
        font-size: 20px;
      }
    }
}

@media only screen and (max-width: 370px) {
    /* For mobile phones: */
    .forecast-container {
      height: 89vh;
    }

    .location {
      font-size: 32px;
    }

    .empty-state {
      .empty_state__message {
        font-size: 22px;
        padding: 0 $s-size;
      }

      .empty-state__drake {
        font-size: 18px;
      }
    }
}
