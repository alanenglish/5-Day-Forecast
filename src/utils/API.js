// const SERVER_URL = '/api';
const FORECASTS_URL = 'https://cors-anywhere.herokuapp.com/https://www.metaweather.com/api/location/';

// export default async function fetchForecasts(woe) {
//   const response = await fetch(`${SERVER_URL}/${woe}`);
//   return await response.json();
// }

export default async function fetchForecasts(woe) {
  const response = await fetch(`${FORECASTS_URL}/${woe}`);
  return await response.json();
}

// function for webpack dev-server (heroku proxy)
// async fetchWeatherForecast(woe) {
//   this.setState({ error: '' });
//   const response = await fetchForecasts(woe);
//
//   if (response.detail) {
//     this.setState({
//       error: 'Location not found, please enter a valid WOE ID.',
//       isLoading: false
//     });
//   } else {
//     this.setState({
//       forecasts: response.consolidated_weather.slice(0, 5),
//       city: response.title,
//       state: response.parent.title,
//       isLoading: false
//     });
//   }
// }

// function for build:dev and yarn start (express server proxy)
// async fetchWeatherForecast(woe) {
//   this.setState({ error: '' });
//   const response = await fetchForecasts(woe);

//   if (response.status !== 200) {
//     this.setState({
//       error: response.error,
//       isLoading: false
//     });
//   } else {
//     this.setState({
//       forecasts: response.data.consolidated_weather.slice(0, 5),
//       city: response.data.title
//       state: resonse.data.parent.title
//       isLoading: false
//     });
//   }
// }
