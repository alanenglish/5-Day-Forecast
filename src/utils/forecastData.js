export const forecastData = [
  {
    id: 5120283761442816,
    weather_state_name: 'Light Rain',
    weather_state_abbr: 'lr',
    wind_direction_compass: 'WSW',
    applicable_date: '2018-12-08',
    min_temp: 7.9325,
    max_temp: 12.5,
    the_temp: 11.445,
    wind_speed: 14.321992441891355,
    wind_direction: 250.87525299232726,
    air_pressure: 1006.64,
    humidity: 74,
    visibility: 8.360549391553329,
    predictability: 75
  },
  {
    id: 6511039873024000,
    weather_state_name: 'Showers',
    weather_state_abbr: 's',
    wind_direction_compass: 'WNW',
    created: '2018-12-08T14:56:02.768346Z',
    applicable_date: '2018-12-09',
    min_temp: 6.16,
    max_temp: 10.1225,
    the_temp: 9.684999999999999,
    wind_speed: 11.841569394687028,
    wind_direction: 303.5076993839963,
    air_pressure: 1011.5699999999999,
    humidity: 76,
    visibility: 10.846034160502665,
    predictability: 73
  },
  {
    id: 5976522334666752,
    weather_state_name: 'Heavy Cloud',
    weather_state_abbr: 'hc',
    wind_direction_compass: 'WNW',
    created: '2018-12-08T14:56:02.850056Z',
    applicable_date: '2018-12-10',
    min_temp: 4.395,
    max_temp: 8.3075,
    the_temp: 8.235,
    wind_speed: 6.999857623568361,
    wind_direction: 290.1214136959148,
    air_pressure: 1029.7,
    humidity: 79,
    visibility: 12.329868567565418,
    predictability: 71
  },
  {
    id: 6033439778144256,
    weather_state_name: 'Showers',
    weather_state_abbr: 's',
    wind_direction_compass: 'S',
    created: '2018-12-08T14:56:02.951337Z',
    applicable_date: '2018-12-11',
    min_temp: 5.1,
    max_temp: 10.2125,
    the_temp: 8.95,
    wind_speed: 4.6199723653424005,
    wind_direction: 171.65388231522883,
    air_pressure: 1025.9299999999998,
    humidity: 90,
    visibility: 11.759139127495427,
    predictability: 73
  },
  {
    id: 6529841964777472,
    weather_state_name: 'Showers',
    weather_state_abbr: 's',
    wind_direction_compass: 'SSW',
    created: '2018-12-08T14:56:03.048969Z',
    applicable_date: '2018-12-12',
    min_temp: 6.0175,
    max_temp: 8.8775,
    the_temp: 8.815000000000001,
    wind_speed: 5.574268742977015,
    wind_direction: 204.10530880492453,
    air_pressure: 1020.015,
    humidity: 81,
    visibility: 13.92120019088523,
    predictability: 73
  }
];
