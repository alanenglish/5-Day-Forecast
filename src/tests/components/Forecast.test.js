import React from 'react';
import { shallow } from 'enzyme';
import Forecast from '../../components/Forecast';
import { forecastData } from '../../utils/forecastData';
/* eslint-env jest */

test('should render Forecast component correctly when no forecasts present', () => {
  const wrapper = shallow(<Forecast />);
  expect(wrapper).toMatchSnapshot();
});

test('should render Forecast component correctly when forecasts are present', () => {
  const wrapper = shallow(<Forecast />);
  wrapper.setState({
    forecasts: forecastData,
    city: 'London',
    state: 'England',
    error: '',
    isLoading: false
  });

  expect(wrapper).toMatchSnapshot();
});
