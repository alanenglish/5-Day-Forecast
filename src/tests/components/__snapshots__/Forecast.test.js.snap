// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Forecast component correctly forecasts are present 1`] = `
<div>
  <SearchBar
    onSubmit={[Function]}
  />
  <div
    className="forecast-container "
  >
    <div>
      <div
        className="location"
      >
        London
        , 
        England
      </div>
      <ul
        className="list-inline forecast-list"
      >
        <ForecastDetails
          forecast={
            Object {
              "air_pressure": 1006.64,
              "applicable_date": "2018-12-08",
              "humidity": 74,
              "id": 5120283761442816,
              "max_temp": 12.5,
              "min_temp": 7.9325,
              "predictability": 75,
              "the_temp": 11.445,
              "visibility": 8.360549391553329,
              "weather_state_abbr": "lr",
              "weather_state_name": "Light Rain",
              "wind_direction": 250.87525299232726,
              "wind_direction_compass": "WSW",
              "wind_speed": 14.321992441891355,
            }
          }
          key="5120283761442816"
        />
        <ForecastDetails
          forecast={
            Object {
              "air_pressure": 1011.5699999999999,
              "applicable_date": "2018-12-09",
              "created": "2018-12-08T14:56:02.768346Z",
              "humidity": 76,
              "id": 6511039873024000,
              "max_temp": 10.1225,
              "min_temp": 6.16,
              "predictability": 73,
              "the_temp": 9.684999999999999,
              "visibility": 10.846034160502665,
              "weather_state_abbr": "s",
              "weather_state_name": "Showers",
              "wind_direction": 303.5076993839963,
              "wind_direction_compass": "WNW",
              "wind_speed": 11.841569394687028,
            }
          }
          key="6511039873024000"
        />
        <ForecastDetails
          forecast={
            Object {
              "air_pressure": 1029.7,
              "applicable_date": "2018-12-10",
              "created": "2018-12-08T14:56:02.850056Z",
              "humidity": 79,
              "id": 5976522334666752,
              "max_temp": 8.3075,
              "min_temp": 4.395,
              "predictability": 71,
              "the_temp": 8.235,
              "visibility": 12.329868567565418,
              "weather_state_abbr": "hc",
              "weather_state_name": "Heavy Cloud",
              "wind_direction": 290.1214136959148,
              "wind_direction_compass": "WNW",
              "wind_speed": 6.999857623568361,
            }
          }
          key="5976522334666752"
        />
        <ForecastDetails
          forecast={
            Object {
              "air_pressure": 1025.9299999999998,
              "applicable_date": "2018-12-11",
              "created": "2018-12-08T14:56:02.951337Z",
              "humidity": 90,
              "id": 6033439778144256,
              "max_temp": 10.2125,
              "min_temp": 5.1,
              "predictability": 73,
              "the_temp": 8.95,
              "visibility": 11.759139127495427,
              "weather_state_abbr": "s",
              "weather_state_name": "Showers",
              "wind_direction": 171.65388231522883,
              "wind_direction_compass": "S",
              "wind_speed": 4.6199723653424005,
            }
          }
          key="6033439778144256"
        />
        <ForecastDetails
          forecast={
            Object {
              "air_pressure": 1020.015,
              "applicable_date": "2018-12-12",
              "created": "2018-12-08T14:56:03.048969Z",
              "humidity": 81,
              "id": 6529841964777472,
              "max_temp": 8.8775,
              "min_temp": 6.0175,
              "predictability": 73,
              "the_temp": 8.815000000000001,
              "visibility": 13.92120019088523,
              "weather_state_abbr": "s",
              "weather_state_name": "Showers",
              "wind_direction": 204.10530880492453,
              "wind_direction_compass": "SSW",
              "wind_speed": 5.574268742977015,
            }
          }
          key="6529841964777472"
        />
      </ul>
    </div>
  </div>
</div>
`;

exports[`should render Forecast component correctly when no forecasts present 1`] = `
<div>
  <SearchBar
    onSubmit={[Function]}
  />
  <div
    className="forecast-container empty-state__background-img"
  >
    <div
      className="empty-state"
    >
      <p
        className="empty_state__message"
      >
        Please enter a WOE ID in order to display forecast. Try running through the index for your WOE's.
      </p>
      <em
        className="empty-state__drake"
      >
        -Drake
      </em>
    </div>
  </div>
</div>
`;
